### ISSUE TITLE
Unspecified properties in `rule-properties-order`, order should not matter

### ISSUE THREAD

**Issue by MadLittleMods** [stylelint/stylelint#374]
### [`rule-properties-order`](https://github.com/stylelint/stylelint/blob/master/src/rules/rule-properties-order/README.md)

When a property is not defined, it should be able to go wherever in the order.

```js
postcss([
	stylelint({
		rules: {
			'rule-properties-order': [2, [
					'color'
				]
			]
		}
	})
])
```

Currently, `stylelint` complains about the order; It wants `color` to come first.

```css
.foo {
	border: 1px solid #000;
	color: #f00;
}
```

**Comment by jeddy3** [comment]
Good suggestion. I think the current behaviour is that unspecified properties are expected to come at the bottom of the declaration list, as that was the easiest to implement at the time. The `rules-properties-order` rule could be enhanced a lot. @MadLittleMods Is this something you'd be able to give us a hand with?

@davidtheclark What do you think? Provide an option to override the current behaviour with skipping over unspecified properties instead?

**Comment by MadLittleMods** [comment]
@jeddy3 This issue is actually blocking me from using this rule as it generates way too many extraneous errors. I can try taking at a stab at it tonight.

Where would we fit in this extra option? Just another item in the array?

```
// Leave unspecified alone should be default
[2, ['color', 'border']] = [2, 'noop', ['color', 'border']]
// Keep unspecified properties at the bottom
[2, 'bottom', ['color', 'border']]
// Error about unspecified properties
[2, 'error', ['color', 'border']]
```

*note:* Be sure that this plays nicely with stylelint/stylelint#373

**Comment by jeddy3** [comment]
> I can try taking a stab at it tonight

That'd be awesome!

Each of the rules follow the same signature of having an object of optional options as the third parameter. See the [options validator code within `at-rule-empty-line-before`](https://github.com/stylelint/stylelint/blob/master/src/rules/at-rule-empty-line-before/index.js#L19) as an example. And how the [optional options](https://github.com/stylelint/stylelint/blob/master/src/rules/indentation/README.md#optional-options) for `indentation` work.

So, `rule-properties-order` might go something like:

```js
"alphabetical" | ["array", "of", "properties"], {
  skipUnspecified: true|false
  expandShorthand: true|false
}
```

The default being `false` if we don't want breaking changes. `true` if we're OK with them. Also, we'll probably need better names for the keys :)

How does that sound to you?

We might not want to provide an `"error"` option as that might be stepping on the toes of [`property-white/blacklist`](https://github.com/stylelint/stylelint/blob/master/docs/rules.md#property)?

Before starting though I suggest waiting on feedback from @davidtheclark, as he's more knowledgable about these things than I.

**Comment by jeddy3** [comment]
Just to clarify, if we go with what's probably most useful to the user but will be a breaking change... the options might look something like:

```js
"alphabetical" | ["array", "of", "properties"], {
  unspecifiedAtBottom: true|false,
  doNotExpandShorthand: true|false
}
```

Default being `false` again.

This time around the default behaviour is to skip over unspecified properties and to expand out shorthand. The user then has the option to change this behaviour if so desired.

**Comment by MadLittleMods** [comment]
@jeddy3 Good points on the `error`, already covered by [white/black list rules](https://github.com/stylelint/stylelint/blob/master/docs/rules.md#property).

How about these keys/options?

```
"alphabetical" | ["array", "of", "properties"], {
  unspecified: 'bottom' || 'skip',
  processor: function noop(rules) { return rules; } || 'expand' || function custom() { return {} }
}
```

**Comment by jeddy3** [comment]
> How about these keys/options?

Cool. I think we're getting closer :)

> `unspecified: 'bottom' || 'skip',`

What is the behaviour if no optional option is specified? I think we need to take that into account when deciding on this option key and values.

> `processor: function noop(rules) { return rules; }`

It might be worth sticking with explicit options as, I believe in the future, the config can be json or yaml. Although, I might be misunderstanding the intent of the `processor` option. 

It looks like the use case you had in mind is automatically extrapolating longhand properties, but optionally specifying an extrapolated order for positions.  So, perhaps an option like:

```js
expandPositionOrder: array, // ["left", "top" etc.]
```

Lastly, I'm now wondering whether we're overly complicating things... if a user wants this amount of control over the order, isn't it just as easy for them to specify it all in the primary options array? We'll only be saving 16 of so lines in the config as aren't `padding`, `margin` and `border` the only longhand properties with positions? I'm unsure though.

**Comment by MadLittleMods** [comment]
@jeddy3 In my example, the first variable is the default. So default would be `unspecified: 'bottom'`, and `processor: 'noop'` in order to not break things for people who rely on this weird behaviour.

The `processor` can accept a string: `noop` or `expand-shorthand`. Thinking about it now, if people want to do a custom processor function they can do it themselves outside of the plugin. I think it should be the opposite of what you are thinking about the  `tlbr`(top, left, bottom, right`) order. `expand-shorthand` covers this and if you want something different, just specify.

**Comment by jeddy3** [comment]
> `expand-shorthand` covers this and if you want something different, just specify.

That seems sensible to me :)

**Comment by davidtheclark** [comment]
```
unspecified: 'bottom' || 'skip'
```

I like this structure (keywords), because, as @MadLittleMods pointed out, it allows for other possibilities in the future (e.g. if anybody ever wants "top").

How about we keep `"bottom"` as the default for now until we feel like we should again bump a major version and make breaking changes? At that point I'd say we should go with `"skip"` as default, because I think it seems the least obtrusive.

If & when we make that breaking release, I'd say we should also implement automatic expansion of hyphenated rules, as suggested in https://github.com/stylelint/stylelint/issues/373. I think that makes good sense. I would point out, though, that the simplest implementation would probably to *not* sort by the string after the hyphen, or else to sort alphabetically after the hyphen — if the user wants `border-top` before `before-left`, they can specify that in their list. Because this expansion would also apply to `font-` and `animation-` and other properties that do not have top, left, etc.

I hope to finish an initial standalone tool & cli sometime very soon, so maybe that would be a good opportunity for the breaking release.

How does all that sound?

**Comment by jeddy3** [comment]
> How does all that sound?

All good :)

**Comment by MadLittleMods** [comment]
@davidtheclark `border-top` and `border-left` should work from the `tlbr`(`top`, ...) if specified otherwise alphabetical.

Although I suggest we make some sensible defaults instead. For example people always specify border this way in shorthand even though it doesn't matter.

```
border-width: 3px;
border-style: solid;
border-color: #f00;
```

**Comment by davidtheclark** [comment]
See PR #392
